refactor: use a Map instead of an Object in dom/data

alpadev opened the pull request:
Fixes #32123

Preview: <https://deploy-preview-32180--twbs-bootstrap.netlify.app/>

alpadev commented:
@XhmikosR I realised something in the dom/data spec https://github.com/twbs/bootstrap/blob/main/js/tests/unit/dom/data.spec.js#L30-L44 

The test says 'should change data if something is already stored' but it expects the `bsKey` to be defined. Also if `data.test` is updated, there would be no need to call `setData` again as the object from the first call remains the same.

rohit2sharma95 commented on js/src/dom/data.js:
```diff
@@ -24,7 +24,7 @@ const mapData = (() => {
       }
 
-      storeData[element.bsKey.id] = data
+      storeData.set(element.bsKey.id, data)
```
The `element` can not be used directly here? 🤔 
Ref: https://github.com/twbs/bootstrap/issues/32123#issuecomment-730180864

Johann-S replied:
yep the element should be used here to the data

alpadev replied:
Uhm that's a bit contradictory to this response https://github.com/twbs/bootstrap/issues/32123#issuecomment-728855901 but well

Johann-S replied:
haha yes sorry @alpadev but things can change 🤷

XhmikosR commented:
FYI this results in a nice size decrease 🙂 https://ja2r7.app.goo.gl/5pUhLMyZD1iuZgYe9

I move this to beta1 @Johann-S. Tests do seem to pass, although I notice a few of them are removed.

rohit2sharma95 commented:
Yes, we discussed the tests here: https://github.com/alpadev/bootstrap/pull/1, though
Few of them are removed, I am not sure which tests should be added again 🙂

XhmikosR commented on js/src/alert.js:
```diff
@@ -49,7 +49,7 @@ class Alert {
     this._element = element
 
     if (this._element) {
-      Data.setData(element, DATA_KEY, this)
```
BTW the `DATA_KEY` constants are only used in one place in each of the plugins except for popover.js and tooltip.js. The minifier already inlines those, so I guess we better keep them for consistency.

XhmikosR commented on js/src/dom/data.js:
```diff
@@ -11,57 +11,16 @@
-  setData(instance, key, data) {
-    mapData.set(instance, key, data)
+export default {
+  setData(element, data) {
```
Maybe we should rename the objects to `set`, `get` and `remove` @Johann-S? I mean, we use them as `Data.foo` so having `Data/fooData` seems a little redundant. Just thinking out loud here :)

Johann-S replied:
yep I agree too 👍

Johann-S commented:
LGTM 👍 since we don't allow different instance on an element 👌

XhmikosR commented:
@Johann-S My only concerns are:

1. if the removed tests should all be removed and if we still cover all cases we did before, like passing non`-bs` prefixed data attributes etc
2. I suppose if we want to support multiple keys in the future it shouldn't be a lot of work, right?

All in all, I love this change; it makes the code simpler and also saves us a good amount of code. I just want to make sure we aren't missing anything and that we won't be circling around in the future 🙂

alpadev commented:
If it's expected to have multiple keys/instances on one element in a closer future, this might be not the ideal implementation to go with, I guess.

This basically removes the circuit of mapping a `UID` (and an `DATA_KEY`), that both been stored in an object at the elements' `bsKey`, to an instance that is saved in the `mapData`, and instead uses the 1:1 relationship of element and instance directly with the element being the key to the `mapData`s' instance.

So before it was like:
A `PluginInstance <> mapData.id <> element.bsKey.id` mapping, with the `DATA_KEY` only being some kind of validation to check if `Data.getData` or `Data.removeData` is called with the same `DATA_KEY` that has been stored in `element.bsKey` when `Data.setData` was called (the first time).

Now it's like:
A `PluginInstance <> mapData.element <> element` mapping, with the `DATA_KEY` not being used here at all.

Another thing to mention, is that `Plugin.getInstance(element)` isn't "type-safe" anymore. Before if I had a `Tooltip` bound to an element, `Tooltip.getInstance(element)` would return some Tooltip-Instance, while `Button.getInstance(element)` returned `null`. Now every Plugin.getInstance returns whatever is saved in `mapData` for that element or `null`.

XhmikosR commented:
Sounds like we need to rethink this. Let's keep this on hold for a few days
until we get more feedback

On Sat, Nov 21, 2020, 19:45 alpadev <notifications@github.com> wrote:

> If it's expected to have multiple keys/instances on one element in a
> closer future, this might be not the ideal implementation to go with, I
> guess.
>
> This basically removes the circuit of mapping a UID (and an DATA_KEY),
> that both been stored in an object at the elements' bsKey, to an instance
> that is saved in the mapData, and instead uses the 1:1 relationship of
> element and instance directly with the element being the key to the
> mapDatas' instance.
>
> So before it was like:
> A PluginInstance <> mapData.id <> element.bsKey.id mapping, with the
> DATA_KEY only being some kind of validation to check if Data.getData or
> Data.removeData is called with the same DATA_KEY that has been stored in
> element.bsKey when Data.setData was called (the first time).
>
> Now it's like:
> A PluginInstance <> mapData.element <> element mapping, with the DATA_KEY
> not being used here at all.
>
> Another thing to mention, is that Plugin.getInstance(element) isn't
> "type-safe" anymore. Before if I had a Tooltip bound to an element,
> Tooltip.getInstance(element) would return some Tooltip-Instance, while
> Button.getInstance(element) returned null. Now every Plugin.getInstance
> returns whatever is saved in mapData for that element or null.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twbs/bootstrap/pull/32180#issuecomment-731611595>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACVLNNCPAEPZ42CF2D3JXDSQ74C7ANCNFSM4TZA7NYQ>
> .
>

alpadev commented:
If I was to propose some implementation for the data, it should look something like this:

```js
const elementMap = new Map();

export default {
  set(element, key, instance) {
    if (!elementMap.has(element)) {
      elementMap.set(element, new Map());
    }

    const instanceMap = elementMap.get(element);

    // make it clear we only want one instance per element
    // can be replaced at a later time when multiple key/instances are allowed
    if (!instanceMap.has(key) && instanceMap.size !== 0) {
      throw new Error(
        "Sorry, we currently allow only one instance per element."
      );
    }

    instanceMap.set(key, instance);
  },

  get(element, key) {
    if (elementMap.has(element)) {
      return elementMap.get(element).get(key) || null;
    }

    return null;
  },

  remove(element, key) {
    if (elementMap.has(element)) {
      const instanceMap = elementMap.get(element);

      instanceMap.delete(key);

      // free up references if there are no instances left for an element
      if (!instanceMap.size) {
        elementMap.delete(element);
      }
    }
  }
};
```

This uses both an `element` and a `key` to be used to store some data/instance. It'd also allow for multiple `key<>interface` mappings.

Jbo666 commented:
Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: alpadev <notifications@github.com>
Sent: Saturday, November 21, 2020, 11:49 AM
To: twbs/bootstrap
Cc: Subscribed
Subject: Re: [twbs/bootstrap] refactor: use a Map instead of an Object in dom/data (#32180)


If it's expected to have multiple keys/instances on one element in a closer future, this might be not the ideal implementation to go with, I guess.

This basically removes the circuit of mapping an UID (and a DATA_KEY), that both been stored in an object at the elements' bsKey, to an instance that is saved in the mapData, and instead uses the 1:1 relationship of element and instance directly with the element being the key to the mapDatas' instance.

So before it was like:
A PluginInstance <> mapData.id <> element.bsKey.id mapping, with the DATA_KEY only being some kind of validation to check if Data.getData or Data.removeData is called with the same DATA_KEY that has been stored in element.bsKey when Data.setData was called (for the first time).

Now it's like:
A PluginInstance <> mapData.element <> element mapping, with the DATA_KEY not being used here at all.

Another thing to mention, is that Plugin.getInstance(element) isn't "type-safe" anymore. Before if I had a Tooltip bound to an element, Tooltip.getInstance(element) would return some Tooltip-Instance, while Button.getInstance(element) returned null. Now every Plugin.getInstance returns whatever is saved in mapData for that element or null.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/twbs/bootstrap/pull/32180#issuecomment-731611595>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/APV6C4UUQTIUJG2BPZE4L3TSQ74RHANCNFSM4TZA7NYQ>.

XhmikosR commented:
@alpadev yup, that sounds like a better approach since that seems closer to what we have right now :)

Johann-S commented:
@alpadev said :
> Another thing to mention, is that Plugin.getInstance(element) isn't "type-safe" anymore. Before if I had a Tooltip bound to an element, Tooltip.getInstance(element) would return some Tooltip-Instance, while Button.getInstance(element) returned null. Now every Plugin.getInstance returns whatever is saved in mapData for that element or null if nothing saved.

The old behavior should be kept here, it's important to be type safe.

And I like your other implementation (https://github.com/twbs/bootstrap/pull/32180#issuecomment-731640223)

BTW when I created `Data` object, I tried to be as close as possible to how jQuery store data when we call `.data` on a jQuery object

XhmikosR commented:
So, assuming we follow @alpadev's [suggestion(https://github.com/twbs/bootstrap/pull/32180#issuecomment-731640223), would this still be considered a breaking change? If so, we need to get this done in beta1, otherwise we can move it to beta2.

rohit2sharma95 commented:
It is not a breaking change then IMO 🤔 
Because as per [the comment](https://github.com/twbs/bootstrap/pull/32180#issuecomment-731640223) of @alpadev the old behaviour would be kept here.

Johann-S commented:
Yep I agree with @rohit2sharma95 👍

alpadev commented:
Alright, then. I think, we could've reimplemented the type-safety by adding instance checks, if it was only for that reason tho. 

Then I'm reseting and updating my branch with the latest proposal.

That also means we won't get that nice improvement in bundle size (as most of it results from removing the `KEY_DATA`), I'm sorry @XhmikosR :). Maybe it makes sense to consider some kind of generic plugin base class that gets extended by each plugin. That maybe could allow for some generalization and thus some savings in that regard.

@Johann-S we usually develop solutions to a certain requirement or problem. Like you said, things may change and so eventually the implementation does. 👍

XhmikosR commented:
> Maybe it makes sense to consider some kind of generic plugin base class that gets extended by each plugin. That maybe could allow for some generalization and thus some savings in that regard.

@alpadev yeah, it's already in the TODO list, see #29370, although the size increases, it's not a big increase.

alpadev committed:
refactor: use Map instead of Object in dom/data

alpadev committed:
test: update spec for dom/data

alpadev commented:
I guess, for the spec it would be better to export the map and validate on it directly but my experience in writing tests is rather basic.

alpadev commented on js/src/dom/data.js:
```diff
@@ -11,57 +11,45 @@
-        }
-        id++
-      }
+const elementMap = new WeakMap()
```
Used a WeakMap here. The advantage is, it holds weak references on the elements and thus shouldn't prevent them from getting garbage collected. Not sure if there are any concerns about this.

XhmikosR replied:
Not super familiar with this either. I'd go with a Map personally, but maybe wait for someone else to chime in

rohit2sharma95 replied:
Agree with @alpadev here

> But `WeakMap` automatically also removes the key along with the information kept by that key when the key (as object) becomes unreachable.

So probably in the future when supporting multiple plugin instances it should be revisited because `dispose` method in the plugins sets the element to null. What do you think? 🙂

alpadev replied:
@rohit2sharma95 tbh I'm not completely aware how WeakMaps internally work, but to my understanding, using an object (or element) as a key here doesn't alter the general behaviour of the garbage collection. It rather uses a weak reference to an object that can be freed automatically when the object has no other reference, instead of a strong reference that would prevent an object from being collected until that reference is explicitely removed.

In that case you mention here, when there are multiple plugin instances each referencing the same element. Setting the element to null in one instance (e.g. `this._element = null`) shouldn't have an impact on any other reference to the same element or the element itself. It's not the deleting the element but rather overwriting that single reference to it with null. 

As such the key-element of the WeakMap should stay intact, as this isn't altered by the code.

XhmikosR replied:
I'm still hesitant to using a WeakMap here. I'd rather err on the safe side and go with a Map personally.

rohit2sharma95 commented on js/src/dom/data.js:
```diff
@@ -11,57 +11,45 @@
-        delete storeData[keyProperties.id]
-        delete element.bsKey
-      }
+    // make it clear we only want one instance per element
+    // can be removed at a later time when multiple key/instances are allowed
+    if (!instanceMap.has(key) && instanceMap.size !== 0) {
+      throw new Error(
+        'Sorry, we currently allow only one instance per element.'
+      )
```
It would be better if you use bootstrap here instead of we

XhmikosR replied:
Not sure if we should throw an error. As much as I like this approach, we generally try to not break people's code.

alpadev replied:
@XhmikosR in general I'm with you on this, but I'm not sure what's the best way to deal with this.
Throwing an error has the advantage of making tests fail in case someone was to implement/refactor a plugin while using multiple instances on an element without being aware of it.
From another perspective the old implementation had kind of a bug, as it wouldn't update the `key` properly when used with a different `key` on any additional call to `setData`. That likely to cause some implications down the line, even when used in production code (maybe without even noticing). So an error could be helpful to detect there is something wrong here.
As an alternative to an error, what do you think would be a good way? We could use console.error and return without altering the map, or do we want to show just the information and continue? We could also mimic the old behaviour, but that means we reintroduce that `key` flaw again.

XhmikosR replied:
Just so that we are on the same boat, I don't believe that no feedback is a good approach. That being said, if we are going to throw an error, this change needs to be in beta1 which is soon to be released, after documenting it in migration notes.

Otherwise, we can't land it later, only in a major version bump.

alpadev commented:
Just pushed changes in regard to the received feedback:

It now uses a `Map` instead of a `WeakMap`. Leaking should be prevented by checking for `instanceMap.size === 0` and removing the element.

Instead of throwing an error, it now uses `console.error` with the following message `Notice: The design of Bootstrap currently discourage to use more than one instance per element. Bound instance(s): ${instanceMap.keys().join(', ')}.`, and continues to set additional keys/instances. Or do you rather want to return from there?

rohit2sharma95 commented:
For now, Bootstrap is not supporting multiple instances on a single element though but IMO it is a good addition @XhmikosR

And this feature should not break anything :thinking: @alpadev

alpadev commented:
@rohit2sharma95 Ideally it shouldn't break something. The signature remains unchanged. If someone uses it with multiple instances on one element, he would get a `console.error` with the information about using it inappropriate, while fixing the problem with the key.

Like I mentioned before - Old behaviour: if someone was to use multiple keys on one element, it would only use the first key and overwrite that on additional calls. Now: it uses multiple keys.

It's possible to mimic the old behaviour exactly like it has been, tho. (But imo I'd consider the old behaviour a bit flawed).

rohit2sharma95 commented:
Agree with your comment: https://github.com/twbs/bootstrap/issues/32123#issuecomment-729107917
And thanks for the explanatory demo :smile:

XhmikosR commented:
So, I don't think we should console.error. We should stick to the current behavior, unless we can't find another way. Otherwise, this could be considered a breaking change AFAICT.

alpadev commented:
@XhmikosR Not sure if I agree on this.

I mean officially it's not a supported feature to use multiple instances on one element according to https://github.com/twbs/bootstrap/issues/32123#issuecomment-731020351. If someone is currently using it, I wouldn't consider it even a bad practise to throw an error, since it's against this intention. Neither do I consider it a breaking change - it's not breaking something that is *supported* behaviour.
I agree tho, that since it's not nice to break user code, it might be a better approach to not make things fail, but rather to give the implementor some hint that he might use it in an inappropriate way. That's whats the console.error is doing.

I'd consider this an edge case anyway, since the documentation doesn't mention it. However, using multiple instances can cause things to break (intransparently) with the current behaviour like in this demo https://codepen.io/-alpa-/pen/KKMExwM. That's why I kinda dislike to keep it the way it is.

Enabling multiple keys allows Data.set/Data.get, to work properly with multiple instances (That's what's causing that many tooltips to spawn in the demo). The behaviour with one instance on one element isn't changed here.

Johann-S commented on js/src/dom/data.js:
```diff
@@ -11,57 +11,44 @@
-    mapData.delete(instance, key)
+
+  remove(element, key) {
+    if (elementMap.has(element)) {
```
I'm a big fan of return as early as possible so I would prefer :
```js
if (!elementMap.has(element)) {
  return;
}
```

Johann-S commented:
I'm not a big fan on console.logs in build files, but in this case I think it make sense to let people know why it doesn't work.

Personally I'm still against allowing people to have multiple plugin instances I think we'll have weird behavior and I don't see any valid use cases currently, that's all for my two cents 😄

Johann-S reviewed (approved):
LGTM 👍 and keep the old behavior in a more up to date way 👌 
thanks @alpadev

rohit2sharma95 commented on js/src/dom/data.js:
```diff
@@ -11,57 +11,44 @@
+    // can be removed later when multiple key/instances are fine to be used
+    if (!instanceMap.has(key) && instanceMap.size !== 0) {
+      // eslint-disable-next-line no-console
+      console.error(`Notice: The design of Bootstrap currently discourages to use more than one instance per element. Bound instance(s): ${[...instanceMap.keys()].join(', ')}.`)
```
I get `[object Map Iterator]` instead the keys' names in the console :thinking: 
![image](https://user-images.githubusercontent.com/26922866/108537456-a11c1380-7303-11eb-8bb0-285a520df91d.png)

alpadev replied:
Mh looks like that's a problem of how the code is processed.

In the source there is `[...instanceMap.keys()].join(', ')` with the spread operator taking care of `keys()` returning an iterator object and iterates that automatically. While in the composed script it is replaced with `[].concat(instanceMap.keys()).join(', ')` which concats that iterator object into an array without iterating.

Replacing it with `Array.from()` to solve that problem.

rohit2sharma95 commented on js/tests/unit/dom/data.spec.js:
```diff
@@ -4,128 +4,103 @@ import Data from '../../../src/dom/data'
 
-      expect(div.bsKey).toBeUndefined()
-    })
+    expect(console.error).toHaveBeenCalled()
+    expect(Data.get(div, UNKNOWN_KEY)).toBe(copy)
```
This expectation makes sure that `UNKNOW_KEY` key is now set on the map. So there are multiple instances registered with the same DOM element. (With `TEST_KEY` and `UNKNOW_KEY`)

alpadev replied:
That's indeed true. I'm open for any feedback in that regard or how to deal with this topic.

rohit2sharma95 replied:
Since it is now supposed that multiple instances will not be supported, for now, you can just show an error in the console and then just return from there. Although the instance of the second plugin would not be created properly, there is an error in the console to let the user know about it.

rohit2sharma95 reviewed (commented):
And this PR now supports multiple plugin instances on a single DOM element (But also closes the issue that was before: https://github.com/twbs/bootstrap/issues/32123#issuecomment-729107917).

There should be a `return` statement after `console.error` in the `set` method :thinking: (If it is supposed not to support multiple instances)

XhmikosR committed:
Update .bundlewatch.config.json

XhmikosR reviewed (changes requested):
There are some last comments to address

alpadev commented:
> I'm not a big fan on console.logs in build files, but in this case I think it make sense to let people know why it doesn't work.
> 
> Personally I'm still against allowing people to have multiple plugin instances I think we'll have weird behavior and I don't see any valid use cases currently, that's all for my two cents 😄

I didn't mean to enforce my opinion here 🙂. My intention was just to explain my standpoint and to receive some input.

As I see it we have those options:

Allow the use of multiple components per element
- Just let it happen and have the implementers figure out problems that may occure 
- Allow it but show a message (console.error/warn/log) to give them a hint **<-- this is how it is for now**

Disallow the use of multiple components
- Throw an error describing the problem
- Overwrite the first component but disposing it before (with or without a message)
- Overwrite the first component without cleaning up (with or without a message)
- Ignore the second component (with or without a message)


> There are some last comments to address

Yeah sorry.. I got my chance to try that thing called Covid and been totally mashed potato in my head for the last week or so. I'm on it.

alpadev committed:
fix: use Array.from instead of spread operator for the error message

as mentioned by rohit2sharma95 - https://github.com/twbs/bootstrap/pull/32180#discussion_r579342295

alpadev committed:
refactor: return early from DomData.remove

as requested by Johann-S - https://github.com/twbs/bootstrap/pull/32180#discussion_r579283552

alpadev committed:
fix: return after console.error when using multiple instances

as clarified with rohit2sharma95 - https://github.com/twbs/bootstrap/pull/32180#discussion_r581608166

rohit2sharma95 committed:
Merge branch 'main' into dom-data-map

rohit2sharma95 commented on js/src/dom/data.js:
```diff
@@ -11,57 +11,47 @@
+    // can be removed later when multiple key/instances are fine to be used
+    if (!instanceMap.has(key) && instanceMap.size !== 0) {
+      // eslint-disable-next-line no-console
+      console.error(`Notice: The design of Bootstrap currently discourages to use more than one instance per element. Bound instance(s): ${Array.from(instanceMap.keys()).join(', ')}.`)
```
Now you can omit **s** from the instance(s) since there is only one instance bound.
```suggestion
      console.error(`Notice: The design of Bootstrap currently discourages to use more than one instance per element. Bound instance: ${Array.from(instanceMap.keys())[0]}.`)

```

alpadev replied:
I think we should also change the `discourages` to `disallow` then and maybe remove `currently` as this isn't something that is likely to be addressed in the closer future.

rohit2sharma95 replied:
/CC @XhmikosR

alpadev committed:
fix: optimize error message

as mentioned by rohit2sharma95 - https://github.com/twbs/bootstrap/pull/32180#discussion_r584165530

XhmikosR committed:
Update data.js

XhmikosR committed:
Merge branch 'main' into dom-data-map
